File: backend/src/modules/auth/auth.service.ts
import bcrypt from "bcryptjs";
import crypto from "node:crypto";
import jwt from "jsonwebtoken";
import type { SignOptions } from "jsonwebtoken";
import { env } from "../../config/env.js";
import { AppError } from "../../shared/errors/app-error.js";
import { durationToMilliseconds } from "../../shared/utils/duration.js";
import { userRepository, type UserEntity } from "../users/user.repository.js";
import { authRepository } from "./auth.repository.js";
import type { AccessTokenPayload, AuthSession, AuthUserDto, RefreshTokenPayload } from "./auth.types.js";
import type { LoginRequestBody, RegisterRequestBody } from "./auth.validation.js";

const passwordSaltRounds = 12;

export class AuthService {
  public async register(input: RegisterRequestBody): Promise<AuthSession> {
    const existingUser = await userRepository.findByEmail(input.email);

    if (existingUser) {
      throw new AppError("EMAIL_ALREADY_REGISTERED", "Email is already registered", 409);
    }

    const passwordHash = await bcrypt.hash(input.password, passwordSaltRounds);
    const user = await userRepository.create({
      email: input.email,
      passwordHash,
      displayName: input.displayName
    });

    return this.createSession(user);
  }

  public async login(input: LoginRequestBody): Promise<AuthSession> {
    const user = await userRepository.findByEmail(input.email);

    if (!user) {
      throw new AppError("INVALID_CREDENTIALS", "Invalid email or password", 401);
    }

    const passwordMatches = await bcrypt.compare(input.password, user.passwordHash);

    if (!passwordMatches) {
      throw new AppError("INVALID_CREDENTIALS", "Invalid email or password", 401);
    }

    return this.createSession(user);
  }

  public async refresh(refreshToken: string): Promise<AuthSession> {
    const payload = this.verifyRefreshToken(refreshToken);
    const tokenHash = await this.hashRefreshToken(refreshToken);
    const persistedToken = await authRepository.findActiveRefreshToken(tokenHash);

    if (!persistedToken || persistedToken.id !== payload.tokenId) {
      throw new AppError("INVALID_REFRESH_TOKEN", "Refresh token is invalid or expired", 401);
    }

    await authRepository.revokeRefreshToken(tokenHash);

    const user = await userRepository.findById(payload.sub);

    if (!user) {
      throw new AppError("INVALID_REFRESH_TOKEN", "Refresh token user no longer exists", 401);
    }

    return this.createSession(user);
  }

  public async logout(refreshToken: string): Promise<void> {
    const tokenHash = await this.hashRefreshToken(refreshToken);
    await authRepository.revokeRefreshToken(tokenHash);
  }

  public async getUserFromAccessToken(accessToken: string): Promise<AuthUserDto> {
    const payload = this.verifyAccessToken(accessToken);
    const user = await userRepository.findById(payload.sub);

    if (!user) {
      throw new AppError("AUTHENTICATION_REQUIRED", "Authentication required", 401);
    }

    return this.toUserDto(user);
  }

  private async createSession(user: UserEntity): Promise<AuthSession> {
    const userDto = this.toUserDto(user);
    const accessToken = this.signAccessToken(userDto);
    const refreshTokenRecord = await authRepository.createRefreshToken({
      userId: userDto.id,
      tokenHash: "pending",
      expiresAt: new Date(Date.now() + durationToMilliseconds(env.REFRESH_TOKEN_TTL))
    });

    const refreshToken = this.signRefreshToken(userDto.id, refreshTokenRecord.id);
    const tokenHash = await this.hashRefreshToken(refreshToken);
    refreshTokenRecord.tokenHash = tokenHash;
    await refreshTokenRecord.save();

    return {
      user: userDto,
      accessToken,
      refreshToken
    };
  }

  private toUserDto(user: UserEntity): AuthUserDto {
    return {
      id: user.id,
      email: user.email,
      displayName: user.displayName
    };
  }

  private signAccessToken(user: AuthUserDto): string {
    const payload: AccessTokenPayload = {
      sub: user.id,
      email: user.email,
      displayName: user.displayName
    };

    return jwt.sign(payload, env.JWT_ACCESS_SECRET, {
      expiresIn: env.ACCESS_TOKEN_TTL as SignOptions["expiresIn"]
    });
  }

  private signRefreshToken(userId: string, tokenId: string): string {
    const payload: RefreshTokenPayload = {
      sub: userId,
      tokenId
    };

    return jwt.sign(payload, env.JWT_REFRESH_SECRET, {
      expiresIn: env.REFRESH_TOKEN_TTL as SignOptions["expiresIn"]
    });
  }

  private verifyAccessToken(accessToken: string): AccessTokenPayload {
    try {
      const payload = jwt.verify(accessToken, env.JWT_ACCESS_SECRET);

      if (!this.isAccessTokenPayload(payload)) {
        throw new AppError("INVALID_ACCESS_TOKEN", "Access token is invalid", 401);
      }

      return payload;
    } catch (error) {
      if (error instanceof AppError) {
        throw error;
      }

      throw new AppError("INVALID_ACCESS_TOKEN", "Access token is invalid or expired", 401);
    }
  }

  private verifyRefreshToken(refreshToken: string): RefreshTokenPayload {
    try {
      const payload = jwt.verify(refreshToken, env.JWT_REFRESH_SECRET);

      if (!this.isRefreshTokenPayload(payload)) {
        throw new AppError("INVALID_REFRESH_TOKEN", "Refresh token is invalid", 401);
      }

      return payload;
    } catch (error) {
      if (error instanceof AppError) {
        throw error;
      }

      throw new AppError("INVALID_REFRESH_TOKEN", "Refresh token is invalid or expired", 401);
    }
  }

  private isAccessTokenPayload(payload: unknown): payload is AccessTokenPayload {
    return (
      typeof payload === "object" &&
      payload !== null &&
      "sub" in payload &&
      "email" in payload &&
      "displayName" in payload &&
      typeof payload.sub === "string" &&
      typeof payload.email === "string" &&
      typeof payload.displayName === "string"
    );
  }

  private isRefreshTokenPayload(payload: unknown): payload is RefreshTokenPayload {
    return (
      typeof payload === "object" &&
      payload !== null &&
      "sub" in payload &&
      "tokenId" in payload &&
      typeof payload.sub === "string" &&
      typeof payload.tokenId === "string"
    );
  }

  private hashRefreshToken(refreshToken: string) {
    return crypto.createHmac("sha256", env.JWT_REFRESH_SECRET).update(refreshToken).digest("hex");
  }
}

export const authService = new AuthService();

File: backend/src/main/java/com/liftorium/service/AuthService.java
package com.liftorium.service;

import com.liftorium.config.AppProperties;
import com.liftorium.dto.AuthDtos.AuthSession;
import com.liftorium.dto.AuthDtos.AuthUserDto;
import com.liftorium.dto.AuthDtos.ForgotPasswordRequest;
import com.liftorium.dto.AuthDtos.LoginRequest;
import com.liftorium.dto.AuthDtos.RegisterInitiateRequest;
import com.liftorium.dto.AuthDtos.RegisterRequest;
import com.liftorium.dto.AuthDtos.RegisterVerifyRequest;
import com.liftorium.dto.AuthDtos.ResetPasswordRequest;
import com.liftorium.entity.PasswordResetRequest;
import com.liftorium.entity.PendingRegistration;
import com.liftorium.entity.RefreshToken;
import com.liftorium.entity.User;
import com.liftorium.exception.AppException;
import com.liftorium.repository.PasswordResetRequestRepository;
import com.liftorium.repository.PendingRegistrationRepository;
import com.liftorium.repository.RefreshTokenRepository;
import com.liftorium.repository.UserRepository;
import io.jsonwebtoken.Claims;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.bson.types.ObjectId;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

  private final UserRepository userRepository;
  private final RefreshTokenRepository refreshTokenRepository;
  private final PendingRegistrationRepository pendingRegistrationRepository;
  private final PasswordResetRequestRepository passwordResetRequestRepository;
  private final PasswordEncoder passwordEncoder;
  private final JwtService jwtService;
  private final OtpService otpService;
  private final EmailService emailService;
  private final AppProperties appProperties;

  public void initiateRegistration(RegisterInitiateRequest input) {
    String email = normalizeEmail(input.email());
    log.info("Registration initiation requested for email: {}", email);

    if (userRepository.existsByEmail(email)) {
      log.warn("Registration initiation failed - email already registered: {}", email);
      throw new AppException("EMAIL_ALREADY_REGISTERED", "Email is already registered", HttpStatus.CONFLICT);
    }

    PendingRegistration pending = pendingRegistrationRepository.findByEmail(email)
        .orElse(PendingRegistration.builder().email(email).attemptCount(0).build());

    Instant rateLimitWindow = Instant.now().minus(
        appProperties.otp().rateLimitWindowMinutes(), ChronoUnit.MINUTES);
    if (pending.getLastAttemptAt() != null
        && pending.getLastAttemptAt().isAfter(rateLimitWindow)
        && pending.getAttemptCount() >= appProperties.otp().maxAttemptsPerWindow()) {
      log.warn("Registration initiation rate limited for email: {}. Attempts: {}", email, pending.getAttemptCount());
      throw new AppException("OTP_RATE_LIMITED", "Too many attempts. Please try again later.", HttpStatus.TOO_MANY_REQUESTS);
    }

    if (pending.getLastAttemptAt() == null || pending.getLastAttemptAt().isBefore(rateLimitWindow)) {
      pending.setAttemptCount(0);
      log.debug("Reset attempt count for email: {}", email);
    }

    String otp = otpService.generateOtp();

    pending.setDisplayName(input.displayName().trim());
    pending.setPasswordHash(passwordEncoder.encode(input.password()));
    pending.setOtpHash(otpService.hashOtp(otp));
    pending.setAttemptCount(pending.getAttemptCount() + 1);
    pending.setLastAttemptAt(Instant.now());
    pending.setExpiresAt(Instant.now().plus(appProperties.otp().expiryMinutes(), ChronoUnit.MINUTES));

    pendingRegistrationRepository.save(pending);
    log.info("Pending registration saved for email: {}. Attempt count: {}", email, pending.getAttemptCount());

    emailService.sendOtp(email, otp);
  }

  public AuthSession verifyRegistration(RegisterVerifyRequest input) {
    String email = normalizeEmail(input.email());
    log.info("Registration verification requested for email: {}", email);

    PendingRegistration pending = pendingRegistrationRepository.findByEmail(email)
        .orElseThrow(() -> {
          log.warn("Registration verification failed - no pending registration found for email: {}", email);
          return new AppException("OTP_EXPIRED", "Verification code has expired. Please request a new one.", HttpStatus.BAD_REQUEST);
        });

    if (!otpService.verifyOtp(input.otp(), pending.getOtpHash())) {
      log.warn("Registration verification failed - invalid OTP for email: {}", email);
      throw new AppException("OTP_INVALID", "Invalid verification code", HttpStatus.BAD_REQUEST);
    }

    if (userRepository.existsByEmail(email)) {
      log.warn("Registration verification failed - email already registered: {}", email);
      pendingRegistrationRepository.deleteByEmail(email);
      throw new AppException("EMAIL_ALREADY_REGISTERED", "Email is already registered", HttpStatus.CONFLICT);
    }

    User user = User.builder()
        .email(email)
        .displayName(pending.getDisplayName())
        .passwordHash(pending.getPasswordHash())
        .build();

    User savedUser = userRepository.save(user);
    log.info("User created successfully via OTP verification. User ID: {}, Email: {}", savedUser.getId(), email);

    pendingRegistrationRepository.deleteByEmail(email);
    log.debug("Pending registration cleaned up for email: {}", email);

    return createSession(savedUser);
  }

  public AuthSession register(RegisterRequest input) {
    String email = normalizeEmail(input.email());

    if (userRepository.existsByEmail(email)) {
      throw new AppException("EMAIL_ALREADY_REGISTERED", "Email is already registered", HttpStatus.CONFLICT);
    }

    User user = User.builder()
        .email(email)
        .displayName(input.displayName().trim())
        .passwordHash(passwordEncoder.encode(input.password()))
        .build();

    return createSession(userRepository.save(user));
  }

  public AuthSession login(LoginRequest input) {
    String email = normalizeEmail(input.email());
    User user = userRepository.findByEmail(email)
        .orElseThrow(() -> new AppException("INVALID_CREDENTIALS", "Invalid email or password", HttpStatus.UNAUTHORIZED));

    if (!passwordEncoder.matches(input.password(), user.getPasswordHash())) {
      throw new AppException("INVALID_CREDENTIALS", "Invalid email or password", HttpStatus.UNAUTHORIZED);
    }

    return createSession(user);
  }

  public AuthSession refresh(String refreshToken) {
    Claims claims = jwtService.verifyRefreshToken(refreshToken);
    String tokenHash = hashRefreshToken(refreshToken);

    RefreshToken persistedToken = refreshTokenRepository
        .findByTokenHashAndRevokedAtIsNullAndExpiresAtAfter(tokenHash, Instant.now())
        .orElseThrow(() -> new AppException("INVALID_REFRESH_TOKEN", "Refresh token is invalid or expired", HttpStatus.UNAUTHORIZED));

    String tokenId = claims.get("tokenId", String.class);
    if (!persistedToken.getId().equals(tokenId)) {
      throw new AppException("INVALID_REFRESH_TOKEN", "Refresh token is invalid or expired", HttpStatus.UNAUTHORIZED);
    }

    persistedToken.setRevokedAt(Instant.now());
    refreshTokenRepository.save(persistedToken);

    User user = userRepository.findById(claims.getSubject())
        .orElseThrow(() -> new AppException("INVALID_REFRESH_TOKEN", "Refresh token user no longer exists", HttpStatus.UNAUTHORIZED));

    return createSession(user);
  }

  public void initiateForgotPassword(ForgotPasswordRequest input) {
    String email = normalizeEmail(input.email());
    log.info("Password reset requested for email: {}", email);

    // Always return success — never reveal whether email exists
    if (!userRepository.existsByEmail(email)) {
      log.info("Password reset requested for non-existent email (suppressed): {}", email);
      return;
    }

    PasswordResetRequest reset = passwordResetRequestRepository.findByEmail(email)
        .orElse(PasswordResetRequest.builder().email(email).attemptCount(0).build());

    Instant rateLimitWindow = Instant.now().minus(
        appProperties.otp().rateLimitWindowMinutes(), ChronoUnit.MINUTES);
    if (reset.getLastAttemptAt() != null
        && reset.getLastAttemptAt().isAfter(rateLimitWindow)
        && reset.getAttemptCount() >= appProperties.otp().maxAttemptsPerWindow()) {
      log.warn("Password reset rate limited for email: {}. Attempts: {}", email, reset.getAttemptCount());
      // Silently suppress to avoid enumeration via rate-limit errors
      return;
    }

    if (reset.getLastAttemptAt() == null || reset.getLastAttemptAt().isBefore(rateLimitWindow)) {
      reset.setAttemptCount(0);
    }

    String otp = otpService.generateOtp();
    reset.setOtpHash(otpService.hashOtp(otp));
    reset.setAttemptCount(reset.getAttemptCount() + 1);
    reset.setLastAttemptAt(Instant.now());
    reset.setExpiresAt(Instant.now().plus(10, ChronoUnit.MINUTES));

    passwordResetRequestRepository.save(reset);
    emailService.sendPasswordResetOtp(email, otp);
  }

  public AuthSession resetPassword(ResetPasswordRequest input) {
    String email = normalizeEmail(input.email());
    log.info("Password reset verification requested for email: {}", email);

    PasswordResetRequest reset = passwordResetRequestRepository.findByEmail(email)
        .orElseThrow(() -> new AppException("OTP_EXPIRED", "Reset code has expired. Please request a new one.", HttpStatus.BAD_REQUEST));

    if (!otpService.verifyOtp(input.otp(), reset.getOtpHash())) {
      log.warn("Password reset failed - invalid OTP for email: {}", email);
      throw new AppException("OTP_INVALID", "Invalid reset code", HttpStatus.BAD_REQUEST);
    }

    User user = userRepository.findByEmail(email)
        .orElseThrow(() -> new AppException("USER_NOT_FOUND", "Account not found", HttpStatus.NOT_FOUND));

    user.setPasswordHash(passwordEncoder.encode(input.newPassword()));
    userRepository.save(user);

    passwordResetRequestRepository.deleteByEmail(email);
    log.info("Password reset successful for email: {}", email);

    return createSession(user);
  }

  public void logout(String refreshToken) {
    String tokenHash = hashRefreshToken(refreshToken);
    refreshTokenRepository.findByTokenHashAndRevokedAtIsNullAndExpiresAtAfter(tokenHash, Instant.now())
        .ifPresent(token -> {
          token.setRevokedAt(Instant.now());
          refreshTokenRepository.save(token);
        });
  }

  public AuthUserDto toUserDto(User user) {
    return new AuthUserDto(user.getId(), user.getEmail(), user.getDisplayName());
  }

  private AuthSession createSession(User user) {
    AuthUserDto userDto = toUserDto(user);
    String accessToken = jwtService.signAccessToken(userDto);
    String refreshTokenId = new ObjectId().toHexString();
    Instant refreshTokenExpiresAt = Instant.now().plus(jwtService.getRefreshTokenTtl());
    String refreshToken = jwtService.signRefreshToken(user.getId(), refreshTokenId);

    refreshTokenRepository.save(RefreshToken.builder()
        .id(refreshTokenId)
        .userId(user.getId())
        .tokenHash(hashRefreshToken(refreshToken))
        .expiresAt(refreshTokenExpiresAt)
        .build());

    return new AuthSession(userDto, accessToken, refreshToken);
  }

  private String normalizeEmail(String email) {
    return email.trim().toLowerCase();
  }

  private String hashRefreshToken(String refreshToken) {
    try {
      Mac mac = Mac.getInstance("HmacSHA256");
      SecretKeySpec keySpec = new SecretKeySpec(
          appProperties.jwt().refreshSecret().getBytes(StandardCharsets.UTF_8),
          "HmacSHA256"
      );
      mac.init(keySpec);
      return bytesToHex(mac.doFinal(refreshToken.getBytes(StandardCharsets.UTF_8)));
    } catch (Exception exception) {
      throw new AppException("TOKEN_HASH_FAILED", "Refresh token could not be processed", HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }

  private String bytesToHex(byte[] bytes) {
    StringBuilder builder = new StringBuilder(bytes.length * 2);
    for (byte value : bytes) {
      builder.append(String.format("%02x", value));
    }
    return builder.toString();
  }
}
